sway-lib-std/src/string.sw

```diff
@@ -0,0 +1,306 @@
+library;
+
+use ::bytes::Bytes;
+use ::convert::From;
+use ::option::Option;
+use ::assert::assert;
+
+/// A UTF-8 encoded growable string
+///
+/// WARNING: As this type is meant to be forward compatible with UTF-8, do *not*
+/// add any mutation functionality or unicode input of any kind until `char` is
+/// implemented, codepoints are *not* guaranteed to fall on byte boundaries
+pub struct String {
+    bytes: Bytes,
+}
+
+impl String {
+    /// Returns `Bytes` giving a UTF-8 representation of the string.
+    pub fn as_bytes(self) -> Bytes {
+        self.bytes
+    }
+
+    /// Gets the amount of memory on the heap allocated to the `String`.
+    pub fn capacity(self) -> u64 {
+        self.bytes.capacity()
+    }
+
+    /// Truncates this `String` to a length of zero, removing all contents.
+    pub fn clear(ref mut self) {
+        self.bytes.clear()
+    }
+
+    /// Converts a vector of ASCII encoded bytes to a `String`.
+    ///
+    /// Each byte is to represent a single character, this supports ASCII but it does **not** support Unicode.
+    ///
+    /// # Arguments
+    ///
+    /// * `bytes` - ASCII bytes which will be converted into a `String`.
+    pub fn from_ascii(bytes: Bytes) -> Self {
+        Self {
+            bytes,
+        }
+    }
+
+    /// Returns `true` if the vector contains no bytes.
+    pub fn is_empty(self) -> bool {
+        self.bytes.is_empty()
+    }
+
+    /// Constructs a new instance of the `String` type.
+    pub fn new() -> Self {
+        Self {
+            bytes: Bytes::new(),
+        }
+    }
+
+    /// Constructs a new instance of the `String` type with the specified capacity.
+    ///
+    /// # Arguments
+    ///
+    /// * `capacity` - The specified amount of memory on the heap to be allocated for the `String`.
+    pub fn with_capacity(capacity: u64) -> Self {
+        Self {
+            bytes: Bytes::with_capacity(capacity),
+        }
+    }
+}
+
+impl From<Bytes> for String {
+    fn from(b: Bytes) -> Self {
+        let mut string = Self::new();
+        string.bytes = b;
+        string
+    }
+
+    fn into(self) -> Bytes {
+        self.bytes
+    }
+}
+
+impl AsRawSlice for String {
+    /// Returns a raw slice to all of the elements in the string.
+    fn as_raw_slice(self) -> raw_slice {
+        asm(ptr: (self.bytes.buf.ptr(), self.bytes.len)) { ptr: raw_slice }
+    }
+}
+
+impl From<raw_slice> for String {
+    fn from(slice: raw_slice) -> String {
+        Self {
+            bytes: Bytes::from_raw_slice(slice),
```

Comment:
```suggestion
            bytes: Bytes::from(slice),
```

There have been recent changes to the `Bytes` type conversion in https://github.com/FuelLabs/sway/pull/4714